### PR TITLE
'Speedy' branch

### PR DESCRIPTION
**Pull request by tobes**
Hi,

This pull request does the following:

* move non `i3status` stuff out of `i3status.py` still more to be done.

* hopefully makes the code a little easier to  follow.

* adds a `check_update` helper to see when modules/i3status need updating.

* reduce the general cpu workload.

```
$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
$ time timeout -sHUP 10m  py3status -c ~/.i3/i3status.conf > /dev/null

real	10m0.007s
user	0m7.822s
sys 	0m2.089s
$ git checkout speedy
Switched to branch 'speedy'
Your branch is up-to-date with 'origin/speedy'.
$ time timeout -sHUP 10m  py3status -c ~/.i3/i3status.conf > /dev/null

real	10m0.006s
user	0m4.967s
sys 	0m2.070s
```

Let me know any thoughts on this

### CODE REVIEW

**Comment by tobes** [comment]
I think some more work could knock another 2 seconds off that user time

**Comment by tobes** [comment]
Version 2, slightly more efficient

```
$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
$ time timeout -sHUP 10m  py3status -c ~/.i3/i3status.conf > /dev/null

real	10m0.011s
user	0m7.766s
sys 	0m2.103s
$ git checkout speedy
Switched to branch 'speedy'
$ time timeout -sHUP 10m  py3status -c ~/.i3/i3status.conf > /dev/null

real	10m0.011s
user	0m3.578s
sys 	0m2.026s
```

**Comment by tobes** [comment]
**Version 3**, refactoring including reducing workload slightly

* Rather than the main loop checking which modules need updating modules 'notify' when they have updated.

* i3status modules self contained in a wrapper class I3statusModule.

* Module and I3statusModule expose same methods to the main loop. 

* Better separation of i3status and py3status modules.

* Lock no longer needed.

* i3status modules only trigger updates if content has changed.

* Although relocated the time correction logic for i3status modules is untouched.

```
$ git checkout master
Switched to branch 'master'
$ time timeout -sHUP 10m  py3status -c ~/.i3/i3status.conf > /dev/null

real	10m0.009s
user	0m8.257s
sys 	0m2.221s
$ git checkout speedy
Switched to branch 'speedy'
$ time timeout -sHUP 10m  py3status -c ~/.i3/i3status.conf > /dev/null

real	10m0.010s
user	0m3.961s
sys 	0m2.175s

```

**Comment by tobes** [comment]
Although this is called 'speedy'  The main advantage IMHO is the re-factoring and separating the i3status/py3status modules but treating them in the same way.

File: py3status/i3status.py [hunk 1]
```diff
@@ -10,18 +10,168 @@
...
+    of the Module class.  It also helps encapsulate the auto time updating for
+    `time` and `tztime`.
+    """
+
+    def __init__(self, module_name, py3statusWrapper):
+        self.module_name = module_name
+        self.i3status = py3statusWrapper.i3status_thread
+        self.is_time_module = module_name.split()[0] in ['time', 'tztime']
+        self.item = {}
+        self.py3statusWrapper = py3statusWrapper
```
Comment (ultrabug) [inline py3status/i3status.py]: If I'm not mistaken, we call it `py3_wrapper` everywhere else ? I'm not a fan of those camelCase like stuff.

**Comment by tobes** [comment]
> If I'm not mistaken, we call it py3_wrapper everywhere else ? I'm not a fan of those camelCase like stuff.

done

**Comment by ultrabug** [comment]
I'd like to merge this nice branch but it breaks the group functionality :( I get some ? instead of my usual group disks output... mind checking this out please ?

**Comment by tobes** [comment]
Good spot.  Sorry was due to how the i3status modules are now stored.  I've fixed it.

Really I need to clean things a bit and add some helper functions to prevent this sort of stuff.

Do you want them in this branch or should I wait till things are merged.  I'd rather do it after the merge so that they are clearer to follow.

**Comment by ultrabug** [comment]
> user method weather_yahoo failed ('item')

almost ! :)

**Comment by ultrabug** [comment]
> Do you want them in this branch or should I wait till things are merged. I'd rather do it after the merge so that they are clearer to follow.

I agree, let's just finish testing/fixing this at it is

**Comment by tobes** [comment]
> user method weather_yahoo failed ('item')

This is unrelated to the changes it fails on 2.9 too.

The json returned has no `item` not sure if this is a temporary change or just broken currently.  But anyway it is the module not these changes so I think it should be fixed independently.

**Comment by ultrabug** [comment]
> This is unrelated to the changes it fails on 2.9 too.

Absolutely sorry, it's just that now that we have Exceptions reported on nagbar (thanks to you) *I see things* ;)

File: py3status/module.py [hunk 2]
```diff
@@ -80,6 +82,17 @@ def clear_cache(self):
...
+    def set_updated(self):
+        """
+        Mark the module as updated
+        """
+        self.py3_wrapper.notify_update(self.module_full_name)
+
+    def get_latest(self):
+        for method in self.methods.values():
+            output = method['last_output']
+        return output
```
Comment (ultrabug) [inline py3status/module.py]: does this mean that if a module has multiple methods, we will output only the latest one ? (I guess we're missing a docstring here)

Comment (tobes) [inline py3status/module.py]: ahh,

I thought it did that but looking at the code it looks like modules can have multiple responses showing. I'll make the changes for that when I get the chance.

You are very thorough with you code reviews, which is great!

Comment (tobes) [inline py3status/module.py]: I've only seen this in wwan_status which I've never used.  but we should keep working as we did

**Comment by tobes** [comment]
OK, we now deal with modules that have multiple outputs.

**Comment by ultrabug** [comment]
Yet another great improvement by @tobes ! :+1: **thank you !**

**Pull request merged**